controllers/secret_controller.go

```diff
@@ -0,0 +1,370 @@
+	}
+
+	// Update the ReplicatedSecret's Status
+	err = localClient.Status().Patch(ctx, rsec, patch)
+	if err != nil {
+		logger.Error(err, "Failed to update replicated secret last transition time", "ReplicatedSecret", req.NamespacedName)
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	// If any cluster had failed state, retry
+	for _, cond := range rsec.Status.Conditions {
+		if cond.Status == corev1.ConditionFalse {
+			return ctrl.Result{Requeue: true}, fmt.Errorf("replication failed")
+		}
+	}
+	return ctrl.Result{}, err
+}
+
+func requiresUpdate(source, dest client.Object) bool {
+	if source.GetGeneration() != 0 {
+		// Compare using generation
+		return source.GetGeneration() > dest.GetGeneration()
+	} else {
+		if srcHash, found := source.GetAnnotations()[api.ResourceHashAnnotation]; found {
+			destHash := dest.GetAnnotations()[api.ResourceHashAnnotation]
+			return srcHash != destHash
+		}
+		// Verify data wasn't manually updated without hash changes..
+		if sourceSec, valid := source.(*corev1.Secret); valid {
+			if destSec, valid := dest.(*corev1.Secret); valid {
+				return !reflect.DeepEqual(sourceSec.Data, destSec.Data)
+			}
+		}
+	}
+	return false
+}
+
+func syncSecrets(src, dest *corev1.Secret) {
+	origMeta := dest.ObjectMeta
+	src.DeepCopyInto(dest)
+	dest.ObjectMeta = origMeta
+	dest.OwnerReferences = []metav1.OwnerReference{}
+
+	// sync annotations, src is more important
+	if dest.GetAnnotations() == nil {
+		dest.Annotations = make(map[string]string)
+	}
+
+	for k, v := range src.Annotations {
+		dest.Annotations[k] = v
+	}
+
+	// sync labels, src is more important
+	if dest.GetLabels() == nil {
+		dest.Labels = make(map[string]string)
+	}
+
+	for k, v := range src.Labels {
+		dest.Labels[k] = v
+	}
+}
+
+func (s *SecretSyncController) verifyHashAnnotation(ctx context.Context, sec *corev1.Secret) error {
+	hash := utils.DeepHashString(sec.Data)
+	if sec.GetAnnotations() == nil {
+		sec.Annotations = make(map[string]string)
+	}
+	if existingHash, found := sec.GetAnnotations()[api.ResourceHashAnnotation]; !found || (existingHash != hash) {
+		sec.GetAnnotations()[api.ResourceHashAnnotation] = hash
+		return s.ClientCache.GetLocalClient().Update(ctx, sec)
+	}
+	return nil
+}
+
+func (s *SecretSyncController) fetchAllMatchingSecrets(namespace string, selector labels.Selector) ([]corev1.Secret, error) {
+	secrets := &corev1.SecretList{}
+	listOption := client.ListOptions{
+		LabelSelector: selector,
+	}
+	err := s.ClientCache.GetLocalClient().List(context.TODO(), secrets, &listOption, client.InNamespace(namespace))
+	if err != nil {
+		return nil, err
+	}
+
+	return secrets.Items, nil
+}
+
+func (s *SecretSyncController) SetupWithManager(mgr ctrl.Manager, clusters []cluster.Cluster) error {
+	err := s.initializeCache()
+	if err != nil {
+		return err
+	}
+
+	// We should only reconcile objects that match the rules
+	toMatchingReplicates := func(secret client.Object) []reconcile.Request {
+		requests := []reconcile.Request{}
```

Comment:
I am not sure I understand how `toMatchingReplicates` works if `s.selectors` is empty. I understand that you first call `s.initializeCache()`, but if there are no ReplicatedSecrets the cache will be empty. @burmanm can you explain how the "bootstrapping" here works?

Reply:
I answered my question :) I see that the mapping function is just for the watch on secrets. I also see where in the `Reconcile` method that the `selectors` caches is updated. 👍

Reply:
toMatchingReplicates will return nothing if the s.selectors is empty. Thus, it will not map any Secret -> ReplicatedSecret. However, we still listen to ReplicatedSecrets (For(&api.ReplicatedSecret{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).) so when a new ReplicatedSecret is created, we add it to the s.selectors (line 150). And then the toMatchingReplicates will see changes that are monitored by that ReplicatedSecret.